pkg/proxy/service.go

```diff
@@ -244,39 +244,45 @@ func (sct *ServiceChangeTracker) newBaseServiceInfo(port *v1.ServicePort, servic
 	return info
 }
 
-type makeServicePortFunc func(*v1.ServicePort, *v1.Service, *BaseServicePortInfo) ServicePort
+type makeServicePortFunc[S ServicePort] func(*v1.ServicePort, *v1.Service, *BaseServicePortInfo) S
+
+// NewBaseServicePortInfo can be used as a makeServicePortFunc for backends that do not
+// need to store backend-specific data
+func NewBaseServicePortInfo(_ *v1.ServicePort, _ *v1.Service, sp *BaseServicePortInfo) *BaseServicePortInfo {
+	return sp
+}
 
 // This handler is invoked by the apply function on every change. This function should not modify the
 // ServicePortMap's but just use the changes for any Proxier specific cleanup.
-type processServiceMapChangeFunc func(previous, current ServicePortMap)
+type processServiceMapChangeFunc[S ServicePort] func(previous, current ServicePortMap[S])
```

Comment:
That won't work; `type processServiceMapChangeFunc[S ServicePort] ...` says that `S` has to implement the `ServicePort` interface, but `servicePortInfo` doesn't implement that; only `*servicePortInfo` does.

We would have to do something like

```golang
type servicePortInfoStruct struct { ... }
type servicePortInfo *servicePortInfoStruct

... proxy.ServicePortMap[servicePortInfo] ...
```

which feels wrong because we don't normally typedef pointer types.

>  In short optimize for reading.

Yeah, that's basically my worry; that this PR doesn't do that at all.

Reply:
OK, ignore my comment about `*`.

WRT optimization for reading, I think the individual proxier code is read more often than the framework code, BUT.... as this moves towards `libproxier` state, there will be more eyes on it.

I don't mind using templates for this - it provides safety as more people use the library, but ultimately this is your call.  In my C++ days, any sort of runtime up-cast was suspect.  I'm happy to see those go away.